libdnf/base/goal.cpp

```diff
@@ -596,10 +604,136 @@ void Goal::Impl::add_rpms_to_goal(base::Transaction & transaction) {
+                        ids_nevra_installed.push_back(id);
+                    }
+                }
+                rpm_goal.add_install(
+                    ids_nevra_installed,
+                    settings.resolve_strict(cfg_main),
+                    settings.resolve_best(cfg_main),
+                    settings.resolve_clean_requirements_on_remove());
+            } break;
             case GoalAction::UPGRADE: {
+                solv::IdQueue ids_upgrades;
+                for (auto id : ids) {
+                    rpm::PackageQuery query(installed);
+                    query.filter_name({pool.get_name(id)});
+                    if (query.empty()) {
+                        // Report when package with the same name is not installed
+                        transaction.p_impl->add_resolve_log(
+                            GoalAction::UPGRADE,
+                            GoalProblem::NOT_INSTALLED,
+                            settings,
+                            {pool.get_nevra(id)},
+                            {},
+                            false);
+                    } else {
+                        query.filter_arch({pool.get_arch(id)});
+                        if (query.empty()) {
+                            // Report when package with the same name is installed for a different architecture
+                            transaction.p_impl->add_resolve_log(
+                                GoalAction::UPGRADE,
+                                GoalProblem::NOT_INSTALLED_FOR_ARCHITECTURE,
+                                settings,
+                                {pool.get_nevra(id)},
+                                {},
+                                false);
+                        } else {
+                            rpm::PackageQuery query_id(base, rpm::PackageQuery::ExcludeFlags::IGNORE_EXCLUDES, true);
+                            query_id.add(rpm::Package(base, rpm::PackageId(id)));
+                            query_id.filter_upgrades();
+                            if (query_id.empty()) {
+                                // Report when package with lower version is not installed
+                                std::string name_arch(pool.get_name(id));
+                                name_arch.append(".");
+                                name_arch.append(pool.get_arch(id));
+                                transaction.p_impl->add_resolve_log(
+                                    GoalAction::UPGRADE,
+                                    GoalProblem::ALREADY_INSTALLED,
+                                    settings,
+                                    {pool.get_nevra(id)},
+                                    {name_arch},
+                                    false);
+                            } else {
+                                // Only installed packages with same name, architecture and lower version can be upgraded
+                                ids_upgrades.push_back(id);
```

Comment:
Alternatively we can add installed all packages with same name - result of query in line 636 `query.filter_name({pool.get_name(id)});` to prevent package downgrade. What do you think?

Reply:
According our discussion I added TODO.

Reply:
Ok. According to the next discussion, I removed TODO and included installed packages with higher or equal version into transaction to prevent downgrade.